### ISSUE TITLE
cannot use a custom secret for Kiali identity

### ISSUE THREAD

**Issue by Vineeth-varma** [kiali/kiali#4761]
Hi Team,

I am Using Kiali-Operator Helm charts. I want https enabled for Kiali CR. So I have created a Secret and mounted using the below fields.
```
spec:
    identity:
      cert_file: /kiali-certificate/tls.crt
      private_key_file: /kiali-certificate/tls.key
    installation_tag: "Vineeth Kiali Dashboard"
    istio_namespace: "istio-system"
    deployment:
      replicas: 1
      custom_secrets:
      - name: "kiali-cert"
        mount: "/kiali-certificate"
        optional: true
```
The above configuration when I was doing once kiali is deployed I was not facing any issues.
**Issue:**
But right at the creation time, When I was using the same configuration I was getting an issue from the operator that it is not able to deploy Kiali CR because of path issue.

Can anyone help me to fix this issue.
Thanks in advance.

**Comment by jmazzitelli** [comment]
Please attach the operator logs or provide any kind of error message that you see.

**Comment by Vineeth-varma** [comment]
![image](https://user-images.githubusercontent.com/60296821/155932853-e8e5d96d-0a9c-4f89-b83b-6dfb5ebaa5ca.png)

Probably this is the issue, once look at it @jmazzitelli

Thanks

**Comment by jmazzitelli** [comment]
Ok, it is pretty obvious what the problem is now. You are declaring a custom secret with the same mount path that the operator is creating/using. You can't mount custom secrets to paths that the operator wants to declare. There are [several of these required paths](https://github.com/kiali/kiali-operator/blob/master/roles/default/kiali-deploy/templates/kubernetes/deployment.yaml#L104-L112) that you cannot use.

I will remove the "bug" label on this issue and make it a "doc" issue - we will put [a blurb in the docs](https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.deployment.custom_secrets[*].mount) to indicate you cannot override a mount path that the operator is using when deploying custom secrets.

To solve your problem, I think all you need to do is change the path. You may have to change the name of the secret, but I'm not sure if that is required. So try this first:

```
identity:
  cert_file: /my-custom-kiali-certificate/tls.crt
  private_key_file: /my-custom-kiali-certificate/tls.key
custom_secrets:
- name: "kiali-cert"
  mount: "/my-custom-kiali-certificate"
  optional: true
```

If that doesn't work due to the name of the secret, change `-name: "kiali-cert` to `-name: "my-custom-kiali-cert"` (and of course rename your secret to `my-custom-kiali-cert`) and that should work.

**Comment by jmazzitelli** [comment]
For the doc fix, we need to do the following. This fix to the docs will show up [here](https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.deployment.custom_secrets[*].mount). We need to change [this](https://github.com/kiali/kiali-operator/blob/v1.47.0/crd-docs/crd/kiali.io_kialis.yaml#L264) to add the following:

```
The custom secret cannot be mounted on a path that the operator will use to mount its secrets.
Make sure you set your custom secret mount path to a unique, unused path. Paths such as
`/kiali-configuration`, `/kiali-cert`, `/kiali-cabundle`, and `/kiali-secret` must not be used
as mount paths for custom secrets.
```

**Comment by Vineeth-varma** [comment]
Yes correct but I am not using any of the above mentioned paths right.

- I did as you said above I have changed the path from /kiali-certificate to /my-custom-kiali-certificate it is throwing me the same error.
- And I tried the other way as you said like changing the secret name "kiali-cert" to "my-custom-kiali-cert" it is throwing me this error.
```
Unable to attach or mount volumes: unmounted volumes=[kiali-cert], unattached volumes=[kiali-configuration kiali-cert kiali-secret kiali-cabundle my-custom-kiali-cert kiali-service-account-token-lgp29]: timed out waiting for the condition
MountVolume.SetUp failed for volume "kiali-cert" : secret "istio.kiali-service-account" not found
Unable to attach or mount volumes: unmounted volumes=[kiali-cert], unattached volumes=[kiali-service-account-token-lgp29 kiali-configuration kiali-cert kiali-secret kiali-cabundle my-custom-kiali-cert]: timed out waiting for the condition
```
Probably we can not change the secret name to something else other than "kiali-cert".

Please check

**Comment by jmazzitelli** [comment]
OK. I will look at this to confirm what is going on.

**Comment by Vineeth-varma** [comment]
Thank you @jmazzitelli I will wait for your update.

**Comment by jmazzitelli** [comment]
There seems to be a problem when attempting to create a custom secret to be used for the kiali identity. Here's what I did to replicate.

First, I created a simple secret. I did this by first creating a tls cert and key via `openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout tls.key -out tls.crt -subj "/CN=foo.bar.com"` and then I base64 encoded the tls.key and tls.crt content and created the k8s secret:
```bash
$ cat <<EOM | kubectl apply -f -
apiVersion: v1
data:
  tls.crt: $(cat tls.crt | base64 -w0)
  tls.key: $(cat tls.key | base64 -w0)
kind: Secret
metadata:
  name: my-secret
  namespace: istio-system
type: kubernetes.io/tls
EOM
```

This creates a secret in the istio-system namespace:
```
$ kubectl get secret my-secret -n istio-system
NAME        TYPE                DATA   AGE
my-secret   kubernetes.io/tls   2      9s
```

I now want to use that (as opposed to the auto-generated secret) for the kiali identity. So I create a Kiali CR with this:

```yaml
deployment:
  custom_secrets:
  - name: "my-secret"
    mount: "/my-secret"
    optional: false
identity:
  cert_file: "/my-secret/tls.crt"
  private_key_file: "/my-secret/tls.key"
```

At this point, the Kiali operator successfully deploys everything. However, the Kiali pod fails to start. Here's the events:

```
$ kubectl describe pod -n istio-system -l app.kubernetes.io/name=kiali
...
Events:
  Type     Reason       Age                  From               Message
  ----     ------       ----                 ----               -------
  Normal   Scheduled    2m24s                default-scheduler  Successfully assigned istio-system/kiali-5545778b4b-m85fw to ci
  Warning  FailedMount  22s                  kubelet            Unable to attach or mount volumes: unmounted volumes=[kiali-cert], unattached volumes=[kiali-configuration kiali-cert kiali-secret kiali-cabundle my-secret kube-api-access-2p4cl]: timed out waiting for the condition
  Warning  FailedMount  17s (x9 over 2m24s)  kubelet            MountVolume.SetUp failed for volume "kiali-cert" : secret "istio.kiali-service-account" not found
```

So it seems the Kiali Operator is attempting to mount kiali-cert when it should not be. This is done here: https://github.com/kiali/kiali-operator/blob/v1.47.0/roles/default/kiali-deploy/templates/kubernetes/deployment.yaml#L127-L132

This is an interesting problem. This appears to be a legacy issue. In the past, Istio used to be able to auto-generate the secret for us, but it no longer does (and we needed that capability when on non-OpenShift clusters; OpenShift gives us the ability to ask for an auto-generated secret). I haven't studied this closer to know how to fix this, but it seems like (at least when on non-OpenShift) we need to NOT attempt to mount any secret (whether optional or not).

I think the workaround is to create a secret called `istio.kiali-service-account` (where `kiali` is the instance name; `kiali` is the default) and do not specify it as a custom_secret (but do specify the `identity` section). I will test this now and report back.

**Comment by jmazzitelli** [comment]
Yes, as I suspected, the workaround will work. @Vineeth-varma , to get this to work, take out the `custom_secrets` section of your Kiali CR, create your secret with the name `istio.kiali-service-account` and make sure the settings in the `identity` section of the Kiali CR uses the `/kiali-cert` mount path. When I did this, Kiali was able to start up successfully:
```
$ kubectl get pod -n istio-system -l app.kubernetes.io/name=kiali
NAME                     READY   STATUS    RESTARTS   AGE
kiali-6ff755877c-lm754   1/1     Running   0          37s
```

My secret:
```
$ kubectl get secret -n istio-system istio.kiali-service-account 
NAME                          TYPE                DATA   AGE
istio.kiali-service-account   kubernetes.io/tls   2      7m51s
```

My Kiali CR `identity` section (note: I do NOT have a `custom_secrets` section):
```
  identity:
    cert_file: /kiali-cert/tls.crt
    private_key_file: /kiali-cert/tls.key
```

**Comment by Vineeth-varma** [comment]
Okay let me try tomorrow and will update here, Thanks for the response and your time. @jmazzitelli

**Comment by jmazzitelli** [comment]
I will leave this open and listed as a "bug" - we should fix this. You should be able to use a `custom_secret` for the identity of Kiali. We should not expect the name of the secret to be hardcoded to `istio.kiali-service-account`.

**Comment by jshaughn** [comment]
adding to sprint as I think this can affect downstream.

**Comment by jmazzitelli** [comment]
@Vineeth-varma this is fixed in PR https://github.com/kiali/kiali-operator/pull/497 - if you know how to build and test Kiali PRs, you can try it and see if it works for you. Otherwise, hopefully this will be available in the next sprint release in a couple weeks.

**Comment by Vineeth-varma** [comment]
You said it is fixed here https://github.com/kiali/kiali-operator/pull/497 but here it is showing the same like how we did before and got into issues. 
And may I know what are kiali PR's ? 
One more thing like why the **Security Context** fields are not there in Kiali CR? How to run kiali with other user? @jmazzitelli

**Comment by jmazzitelli** [comment]
>You said it is fixed here kiali/kiali-operator#497 but here it is showing the same like
> how we did before and got into issues. And may I know what are kiali PR's ?

@Vineeth-varma from your last questions here it sounds like you are not familiar with how to develop/build within a github project (Kiali PRs are just requests to make changes to the source code, which you must then build on your local machine). For now, you can ignore my suggestion that you build and test this Kiali PR. You can wait for the next release of Kiali and this fix for you should be there for you to test.

> One more thing like why the Security Context fields are not there in Kiali CR? How to run kiali with other user? @jmazzitelli

As for Security Contexts and "how to run kiali with other user?", this is an addition question not relevant to this issue and related PR. If you have additional separate questions such as how to provide access to different users, please read the docs here - there are several places in the kiali.io docs that should help:
* https://kiali.io/docs/configuration/rbac/
* https://kiali.io/docs/configuration/authentication/
* https://kiali.io/docs/faq/authentication/

If you have further questions that the docs do not answer, you can search the [Q&A discussion forum here](https://github.com/kiali/kiali/discussions) - many questions and answers about users/RBAC are there. Or you can [create your own question here](https://github.com/kiali/kiali/discussions/new) and someone will answer your question there.
